wyzesense2mqtt.py

```diff
@@ -1,229 +1,392 @@
-    _LOGGER = logging.getLogger("wyzesense2mqtt")
-
-def findDongle():
-    df = subprocess.check_output(["ls", "-la", "/sys/class/hidraw"]).decode("utf-8").lower()
-    for l in df.split("\n"):
-        if ("e024" in l and "1a86" in l):
-            for w in l.split(" "):
-                if ("hidraw" in w):
-                    return "/dev/%s" % w
-
-def on_connect(client, userdata, flags, rc):
-    _LOGGER.info("Connected with result code {0}".format(str(rc)))
-    client.subscribe([(SCAN_TOPIC, MQTT_QOS), (REMOVE_TOPIC, MQTT_QOS)])
-    client.message_callback_add(SCAN_TOPIC, on_message_scan)
-    client.message_callback_add(REMOVE_TOPIC, on_message_remove)
-
-def on_disconnect(client, userdata, rc):
-    _LOGGER.info("Disconnected with result code {0}".format(str(rc)))
-    client.message_callback_remove(SCAN_TOPIC)
-    client.message_callback_remove(REMOVE_TOPIC)
-
-# Process messages
-def on_message(client, userdata, msg):
-    _LOGGER.info("{0} {1}".format(msg.topic, str(msg.payload)))
-
-# Process message to scan for new devices
-def on_message_scan(client, userdata, msg):
-    _LOGGER.info("In on_message_scan: {0}".format(msg.payload.decode()))
-    prescan_result = ws.List()
-    _LOGGER.debug("Result of prescan {0}".format(prescan_result))   
-
-    ws.Scan()
-
-    postscan_result = ws.List()
-    _LOGGER.debug("Result of postscan {0}".format(postscan_result))  
-
-    s = diff(postscan_result, prescan_result)  
-    _LOGGER.info("Diff is {0}".format(s))  
-
-    if s != []:
-        jsonData = json.dumps({'macs': s})
-        _LOGGER.debug(jsonData)
-        client.publish(SCAN_RESULT_TOPIC, payload = jsonData)
+    global LOGGER
+    logging_config = read_yaml_file(LOGGING_CONFIG_FILE)
+    log_path = os.path.dirname(logging_config['handlers']['file']['filename'])
+    try:
+        if (not os.path.exists(log_path)):
+            os.makedirs(log_path)
+    except IOError:
+        print("Unable to create log folder")
+    logging.config.dictConfig(logging_config)
+    LOGGER = logging.getLogger("wyzesense2mqtt")
+    LOGGER.debug("Logging initialized...")
+
+
+# Initialize configuration
+def init_config():
+    global CONFIG
+    LOGGER.debug("Reading configuration...")
+    CONFIG = read_yaml_file(GENERAL_CONFIG_FILE)
+
+
+# Initialize MQTT client connection
+def init_mqtt_client():
+    global MQTT_CLIENT, CONFIG
+
+    # Configure MQTT Client
+    MQTT_CLIENT = mqtt.Client(
+            client_id=CONFIG['mqtt_client_id'],
+            clean_session=CONFIG['mqtt_clean_session']
+    )
+    MQTT_CLIENT.username_pw_set(
+            username=CONFIG['mqtt_username'],
+            password=CONFIG['mqtt_password']
+    )
+    MQTT_CLIENT.reconnect_delay_set(min_delay=1, max_delay=120)
+    MQTT_CLIENT.on_connect = on_connect
+    MQTT_CLIENT.on_disconnect = on_disconnect
+    MQTT_CLIENT.on_message = on_message
+
+    # Connect to MQTT
+    LOGGER.info(f"Connecting to MQTT host {CONFIG['mqtt_host']}")
+    MQTT_CLIENT.connect(
```

Comment:
The MQTT client stuff is interesting. As best I can tell, the keepalive helps with the connect. But the real connection is made when a loop() function is called. The MQTT library has a special function, at the end of this script, called loop_forever() which includes the retry_first_connection option, but otherwise is designed to wait a period of time, and then reconnect to MQTT to send any waiting payloads, and does this over and over again.

The one catch I saw in some testing is that if there is a failure while in that loop, then it is possible the connection will die and may not be re-established. I tried adding a significant amount of error handling around the rest of the script, so that those situations can't happen and the loop won't be killed. Thus some of the "except TimeoutError: pass" pieces of code around the WyzeSense functions.

Reply:
Looked at this a bit more and have added some additional error logging in the on_connect function which gets the return from the connect. It should now exit if a connection is a failing one. Otherwise the loop_forever will reconnect on a temporary failure to connect. Assuming I'm reading the Paho MQTT docs correctly. Marking resolved with that checkin, but please do review.